src/mscorlib/corefx/System/Globalization/TextInfo.Unix.cs

```diff
@@ -17,46 +20,76 @@ public partial class TextInfo
         internal unsafe TextInfo(CultureData cultureData)
         {
             // TODO: Implement this fully.
-            this.m_cultureData = cultureData;
-            this.m_cultureName = this.m_cultureData.CultureName;
-            this.m_textInfoName = this.m_cultureData.STEXTINFO;
+            m_cultureData = cultureData;
+            m_cultureName = m_cultureData.CultureName;
+            m_textInfoName = m_cultureData.STEXTINFO;
+            m_needsTurkishCasing = NeedsTurkishCasing(this.m_textInfoName);
         }
 
+        [System.Security.SecuritySafeCritical]
         private unsafe string ChangeCase(string s, bool toUpper)
         {
-            Contract.Assert(s != null);              
-            // TODO: Implement this fully.
+            Contract.Assert(s != null);
 
-            StringBuilder sb = new StringBuilder(s.Length);
+            char[] dstBuf = new char[s.Length];
 
-            for (int i = 0; i < s.Length; i++)
+            fixed (char* pSrc = s, pDst = dstBuf)
             {
-                sb.Append(ChangeCaseAscii(s[i], toUpper));
+                ChangeCaseCore(pSrc, s.Length, pDst, dstBuf.Length, toUpper);
             }
 
-            return sb.ToString();
+            return new string(dstBuf);
         }
 
+        [System.Security.SecuritySafeCritical]
         private unsafe char ChangeCase(char c, bool toUpper)
         {
-            // TODO: Implement this fully.
-            return ChangeCaseAscii(c, toUpper);
+            char* pSrc = stackalloc char[1];
+            char* pDst = stackalloc char[1];
+
+            pSrc[0] = c;
+
+            ChangeCaseCore(pSrc, 1, pDst, 1, toUpper);
+
+            return pDst[0];
         }
 
-        // PAL Methods end here.
+        // -----------------------------
+        // ---- PAL layer ends here ----
+        // -----------------------------
 
-        internal static char ChangeCaseAscii(char c, bool toUpper = true)
+        private unsafe void ChangeCaseCore(char* pSrc, int cchSrc, char* pDst, int cchDst, bool toUpper)
         {
-            if (toUpper && c >= 'a' && c <= 'z')
+            if (toUpper)
             {
-                return (char)('A' + (c - 'a'));
+                if (!m_needsTurkishCasing)
```

Comment:
Just curious: why did you choose to write it this way rather than ```if (m_needsTurkishCasing)``` and flipping the branches?